File: katalyze/src/main/java/web/CompressedWebDocument.java
package web;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.OutputStream;
import java.util.zip.GZIPOutputStream;

public class CompressedWebDocument implements WebDocument {
	final String contentType;
	final byte[] compressedContent;

	
	static WebDocument Compress(WebDocument source) {
		if (source.isGzipCompressed()) {
			return source;
		}
		
		return new CompressedWebDocument(source.getContentType(), compress(source));
		
	}
	
	public CompressedWebDocument(String contentType, byte[] compressedContent) {
		this.contentType = contentType;
		this.compressedContent = compressedContent;
	}
	
    private static byte[] compress(WebDocument source){
        ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
        try{
            GZIPOutputStream gzipOutputStream = new GZIPOutputStream(byteArrayOutputStream);
            source.writeContents(gzipOutputStream);
            gzipOutputStream.close();
            byteArrayOutputStream.flush();            
        } catch(IOException e){
        		throw new RuntimeException(e);
        }

        return byteArrayOutputStream.toByteArray();
    }	

	@Override
	public String getContentType() {
		return this.contentType;
	}

	@Override
	public void writeContents(OutputStream target) throws IOException {
		target.write(compressedContent);
	}

	@Override
	public boolean isGzipCompressed() {
		return true;
	}
}

File: katalyze/src/main/kotlin/web/CompressedWebDocument.kt
package web

import java.io.ByteArrayOutputStream
import java.io.IOException
import java.io.OutputStream
import java.util.zip.GZIPOutputStream

class CompressedWebDocument(override val contentType: String, val compressedContent: ByteArray) : WebDocument {
    @Throws(IOException::class)
    override fun writeContents(target: OutputStream) {
        target.write(compressedContent)
    }

    override val isGzipCompressed: Boolean = true

    companion object {
        fun Compress(source: WebDocument): WebDocument {
            if (source.isGzipCompressed) {
                return source
            }

            return CompressedWebDocument(source.contentType, compress(source))
        }

        private fun compress(source: WebDocument): ByteArray {
            val byteArrayOutputStream = ByteArrayOutputStream()
            try {
                val gzipOutputStream = GZIPOutputStream(byteArrayOutputStream)
                source.writeContents(gzipOutputStream)
                gzipOutputStream.close()
                byteArrayOutputStream.flush()
            } catch (e: IOException) {
                throw RuntimeException(e)
            }

            return byteArrayOutputStream.toByteArray()
        }
    }
}